File: curring.cpp
#include <string>
#include <functional>
#include <iostream>

using namespace std::placeholders;

std::string catstr(std::string s1, std::string s2) {
  return s1 + s2;
}

int main() {
  auto greet = std::bind(catstr, "Hello ", _1);
  std::cout << greet("Haskell Curry");
}

File: curring.hs
module Currying where
  
catstr :: String -> String -> String
catstr s s' = s ++ s'

catstr' s = \s' -> s ++ s'

greet :: String -> String
greet = catstr "Hello "

curry :: ((a, b) -> c) -> (a -> b -> c)
curry f a b = f (a, b)

uncurry :: (a -> b -> c) -> ((a, b) -> c)
uncurry f (a, b) = f a b

factorizer :: ((a, b) -> c) -> (a -> (b -> c))
factorizer g = \a -> (\b -> g (a, b))